lib/vector-core/buffers/src/disk_v2/common.rs

```diff
@@ -0,0 +1,4 @@
+// We don't want data files to be bigger than 128MB, but we might end up overshooting slightly.
+pub const DATA_FILE_TARGET_MAX_SIZE: u64 = 128 * 1024 * 1024;
```

Comment:
Why'd we choose 128? I'm not opposed, just curious about the reasoning.

Reply:
Seems like from later comments this is an arbitrary constraint, which works for me. Maybe worth linking this comment up with the one in `disk_v2/mod.rs` for clarity's sake.

Reply:
Yeah, honestly, it was just a random pick.  The entire design is predicated around eventually limiting buffer growth so we can provide invariants around number of expected files on disk at any one time.

I figured that buffering up to multiple terabytes is a _possible_, albeit unlikely, scenario and sort of worked backwards from there.  256MB or 512MB would have been better -- 16k files in a directory is going to kneecap the shell of anyone who dares run `ls -l` -- but we also don't delete or shrink files until fully consumed, so the larger the data file, the more data we're consuming above the configured limit.

Side note, which brings up an interesting point: for the sake of simplicity, we don't try and shrink files as reads are completing, but we do internally track the virtual total buffer size.  That is to say, if you have a single data file fully written to (128MB) but you've only read half way through it, the file is still going to be 128MB, but the reported buffer size is ~64MB.  If the buffer size limit was 128MB, you could manage to successfully write.  Thus, there's always a maximum overhead of `configured total buffer size + target maximum size of one data file`.

This all feeds into the choice of 128MB.  It felt like a small enough overhead that we could call it out in the documentation, but technically speaking, we would exceed the configured total buffer size unless we compensated for the fact that data files are dropped in one fell swoop.

The other option is that we track buffer size faithfully, but this means that writes would block for as long as the currently-being-read file exists, since we would only drop it once it was read in its entirety.  That could be a long time for a slow sink, instead of allowing a trickle of writes which will happen when we maintain a virtual total buffer size based on reader/writer positions.

Reply:
> Yeah, honestly, it was just a random pick. The entire design is predicated around eventually limiting buffer growth so we can provide invariants around number of expected files on disk at any one time.
> 
> I figured that buffering up to multiple terabytes is a _possible_, albeit unlikely, scenario and sort of worked backwards from there. 256MB or 512MB would have been better -- 16k files in a directory is going to kneecap the shell of anyone who dares run `ls -l` -- but we also don't delete or shrink files until fully consumed, so the larger the data file, the more data we're consuming above the configured limit.

Makes perfect sense.

> Side note, which brings up an interesting point: for the sake of simplicity, we don't try and shrink files as reads are completing, but we do internally track the virtual total buffer size. That is to say, if you have a single data file fully written to (128MB) but you've only read half way through it, the file is still going to be 128MB, but the reported buffer size is ~64MB. If the buffer size limit was 128MB, you could manage to successfully write. Thus, there's always a maximum overhead of `configured total buffer size + target maximum size of one data file`.
> 
> This all feeds into the choice of 128MB. It felt like a small enough overhead that we could call it out in the documentation, but technically speaking, we would exceed the configured total buffer size unless we compensated for the fact that data files are dropped in one fell swoop.

Ah ha, yes. Yes, we always round up to the nearest multiple of 128MB. Well worth keeping that value lowish, which 128 is.

> The other option is that we track buffer size faithfully, but this means that writes would block for as long as the currently-being-read file exists, since we would only drop it once it was read in its entirety. That could be a long time for a slow sink, instead of allowing a trickle of writes which will happen when we maintain a virtual total buffer size based on reader/writer positions.

Nah, I like the approach we have presently.

Reply:
Good call on rounding up.  It didn't even occur to me that we could just.... tell users that we will round it up to the next highest multiple of 128MB. 🤯   Internally, we would just limit it to the next multiple _down_, leaving a full 128MB as our overhead.

Much simpler to explain it like that as the maximum expected size.